src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -591,6 +597,28 @@ private void OnImportChanged(object sender, EventArgs e)
         {
             VisualStudioMetadataReference reference = (VisualStudioMetadataReference)sender;
 
+            CancellationTokenSource delayTaskCancellationTokenSource;
+            if (_changedReferencesPendingUpdate.TryGetValue(reference, out delayTaskCancellationTokenSource))
+            {
+                delayTaskCancellationTokenSource.Cancel();
+            }
+
+            delayTaskCancellationTokenSource = new CancellationTokenSource();
+            _changedReferencesPendingUpdate[reference] = delayTaskCancellationTokenSource;
+
+            var task = Task.Delay(TimeSpan.FromSeconds(5), delayTaskCancellationTokenSource.Token)
+                .ContinueWith(
+                    OnImportChangedAfterDelay,
+                    reference,
+                    delayTaskCancellationTokenSource.Token,
+                    TaskContinuationOptions.None,
+                    TaskScheduler.FromCurrentSynchronizationContext());
+        }
+
+        private void OnImportChangedAfterDelay(Task previous, object state)
+        {
+            var reference = (VisualStudioMetadataReference)state;
```

Comment:
Add an explicit assert we're on the UI thread here. This should derive from an affinitized object somewhere.